File: push/src/main/java/com/xiaomi/slim/SlimConnection.java
package com.xiaomi.slim;

import android.text.TextUtils;
import com.google.protobuf.micro.ByteStringMicro;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.push.mpcd.Constants;
import com.xiaomi.push.protobuf.ChannelMessage;
import com.xiaomi.push.service.PushClientsManager;
import com.xiaomi.push.service.PushSlimInboundAction;
import com.xiaomi.push.service.PushSlimInboundPlan;
import com.xiaomi.push.service.PushSlimPingPlan;
import io.github.magisk317.mipush.service.runtime.PushSlimConnectionRuntime;
import io.github.magisk317.mipush.service.runtime.PushSocketConnectionRuntime;
import com.xiaomi.push.service.ServiceConfig;
import com.xiaomi.push.service.XMPushService;
import com.xiaomi.smack.Connection;
import com.xiaomi.smack.ConnectionConfiguration;
import com.xiaomi.smack.SocketConnection;
import com.xiaomi.smack.XMPPException;
import com.xiaomi.smack.packet.Packet;
import com.xiaomi.smack.util.TrafficUtils;
import com.xiaomi.stats.StatsHelper;
import io.github.magisk317.mipush.runtime.PushRuntime;
import java.io.IOException;
import java.util.Iterator;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/slim/SlimConnection.class */
public class SlimConnection extends SocketConnection {
    private byte[] mKey;
    private BlobReader mReader;
    private Thread mReaderThread;
    private BlobWriter mWriter;

    public SlimConnection(XMPushService xMPushService, ConnectionConfiguration connectionConfiguration) {
        super(xMPushService, connectionConfiguration);
    }

    private Blob getPing(boolean z) {
        Ping ping = new Ping();
        if (z) {
            ping.setPacketID("1");
        }
        byte[] bArrRetriveStatsAsByte = StatsHelper.retriveStatsAsByte();
        if (bArrRetriveStatsAsByte != null) {
            ChannelMessage.XMMsgPing xMMsgPing = new ChannelMessage.XMMsgPing();
            xMMsgPing.setStats(ByteStringMicro.copyFrom(bArrRetriveStatsAsByte));
            ping.setPayload(xMMsgPing.toByteArray(), null);
        }
        return ping;
    }

    private void initReaderAndWriter() throws XMPPException {
        try {
            this.mReader = new BlobReader(this.socket.getInputStream(), this);
            this.mWriter = new BlobWriter(this.socket.getOutputStream(), this);
            Thread thread = new Thread("Blob Reader (" + this.connectionCounterValue + Constants.SEPARATOR_RIGHT_PARENTESIS) { // from class: com.xiaomi.slim.SlimConnection.1
                @Override // java.lang.Thread, java.lang.Runnable
                public void run() {
                    try {
                        SlimConnection.this.mReader.start();
                    } catch (Exception e) {
                        SlimConnection.this.notifyConnectionError(9, e);
                    }
                }
            };
            this.mReaderThread = thread;
            thread.start();
        } catch (Exception e) {
            throw new XMPPException("Error to init reader and writer", e);
        }
    }

    @Override // com.xiaomi.smack.SocketConnection, com.xiaomi.smack.Connection
    public void batchSend(Blob[] blobArr) {
        for (Blob blob : blobArr) {
            send(blob);
        }
    }

    @Override // com.xiaomi.smack.SocketConnection, com.xiaomi.smack.Connection
    @Deprecated
    public void batchSendPacket(Packet[] packetArr) {
        for (Packet packet : packetArr) {
            sendPacket(packet);
        }
    }

    @Override // com.xiaomi.smack.Connection
    public void bind(PushClientsManager.ClientLoginInfo clientLoginInfo) throws XMPPException {
        synchronized (this) {
            Binder.bind(clientLoginInfo, getChallenge(), this);
        }
    }

    byte[] getKey() {
        byte[] bArr;
        synchronized (this) {
            if (this.mKey == null && !TextUtils.isEmpty(this.challenge)) {
                this.mKey = PushSocketConnectionRuntime.deriveConnectionKey(this.challenge, ServiceConfig.getDeviceUUID());
                if (this.mKey != null) {
                    PushRuntime.observeChannelEvent(null, "slim_key_derived", "SlimConnection.getKey");
                }
            }
            bArr = this.mKey;
        }
        return bArr;
    }

    @Override // com.xiaomi.smack.SocketConnection
    protected void initConnection() throws XMPPException, IOException {
        synchronized (this) {
            initReaderAndWriter();
            this.mWriter.openStream();
        }
    }

    @Override // com.xiaomi.smack.Connection
    public boolean isBinaryConnection() {
        return true;
    }

    void notifyDataArrived(Blob blob) {
        if (blob == null) {
            return;
        }
        if (blob.hasErr()) {
            MyLog.w("[Slim] RCV blob chid=" + blob.getChannelId() + "; id=" + blob.getPacketID() + "; errCode=" + blob.getErrCode() + "; err=" + blob.getErrStr());
        }
        PushSlimInboundPlan inboundPlan = PushSlimConnectionRuntime.planInboundBlob(blob.getChannelId(), blob.getCmd());
        if (inboundPlan.getAction() == PushSlimInboundAction.PingReceived) {
            MyLog.w("[Slim] RCV ping id=" + blob.getPacketID());
        }
        if (inboundPlan.getShouldUpdateLastReceived()) {
            updateLastReceived();
        }
        if (inboundPlan.getEventAction() != null) {
            PushRuntime.observeChannelEvent(null, inboundPlan.getEventAction(), "SlimConnection.notifyDataArrived");
        }
        if (inboundPlan.getConnectionState() != null) {
            com.xiaomi.xmsf.runtime.PushRuntime.observeConnectionState(inboundPlan.getConnectionState(), "SlimConnection.notifyDataArrived", getHost(), inboundPlan.getConnectionReason(), System.currentTimeMillis());
        }
        if (inboundPlan.getDisconnectReasonCode() != null) {
            notifyConnectionError(inboundPlan.getDisconnectReasonCode().intValue(), null);
        }
        Iterator<Connection.ListenerWrapper> it = this.recvListeners.values().iterator();
        while (it.hasNext()) {
            it.next().notifyListener(blob);
        }
    }

    void notifyDataArrived(Packet packet) {
        if (packet == null) {
            return;
        }
        Iterator<Connection.ListenerWrapper> it = this.recvListeners.values().iterator();
        while (it.hasNext()) {
            it.next().notifyListener(packet);
        }
    }

    @Override // com.xiaomi.smack.Connection
    public void send(Blob blob) {
        BlobWriter blobWriter = this.mWriter;
        if (blobWriter == null) {
            throw new IllegalStateException("the writer is null.");
        }
        try {
            int iWrite = blobWriter.write(blob);
            this.writeAlive = System.currentTimeMillis();
            String packageName = blob.getPackageName();
            if (!TextUtils.isEmpty(packageName)) {
                try {
                    TrafficUtils.distributionTraffic(this.mPushService, packageName, iWrite, false, true, System.currentTimeMillis());
                } catch (Throwable th) {
                    MyLog.e(th);
                }
            }
            Iterator<Connection.ListenerWrapper> it = this.sendListeners.values().iterator();
            while (it.hasNext()) {
                it.next().notifyListener(blob);
            }
        } catch (Exception e) {
            throw new IllegalStateException("failed to send blob", e);
        }
    }

    @Override // com.xiaomi.smack.Connection
    @Deprecated
    public void sendPacket(Packet packet) {
        send(Blob.from(packet, null));
    }

    @Override // com.xiaomi.smack.SocketConnection
    protected void sendPing(boolean z) {
        if (this.mWriter == null) {
            throw new IllegalStateException("The BlobWriter is null.");
        }
        Blob ping = getPing(z);
        PushSlimPingPlan pingPlan = PushSlimConnectionRuntime.planSendPing();
        MyLog.w("[Slim] SND ping id=" + ping.getPacketID());
        PushRuntime.observeChannelEvent(null, pingPlan.getEventAction(), "SlimConnection.sendPing");
        send(ping);
        updateLastSent();
    }

    @Override // com.xiaomi.smack.SocketConnection
    protected void shutdown(int i, Exception exc) {
        synchronized (this) {
            BlobReader blobReader = this.mReader;
            if (blobReader != null) {
                blobReader.shutdown();
                this.mReader = null;
            }
            BlobWriter blobWriter = this.mWriter;
            if (blobWriter != null) {
                try {
                    blobWriter.shutdown();
                } catch (Exception e) {
                    MyLog.e(e);
                }
                this.mWriter = null;
                this.mKey = null;
                super.shutdown(i, exc);
            } else {
                this.mKey = null;
                super.shutdown(i, exc);
            }
        }
    }

    @Override // com.xiaomi.smack.Connection
    public void unbind(String str, String str2) throws XMPPException {
        synchronized (this) {
            Binder.unbind(str, str2, this);
        }
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/slim/SlimConnection.kt
package com.xiaomi.slim

import android.text.TextUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.push.protobuf.ChannelMessage
import com.xiaomi.push.service.*
import com.xiaomi.smack.Connection
import com.xiaomi.smack.ConnectionConfiguration
import com.xiaomi.smack.SocketConnection
import com.xiaomi.smack.XMPPException
import com.xiaomi.smack.packet.Packet
import com.xiaomi.smack.util.TrafficUtils
import com.xiaomi.stats.StatsHelper
import java.io.IOException

/*
 * Stock reference: com.xiaomi.xmsf 7.4.67-C (versionCode 70004067),
 * split-XiaomiServiceFrameworkCN-master.apk sha256 444e9f128591e04e38672bfe44a246ab3fa97ae68e95882839d8a7afe766df2b,
 * JADX path: com.xiaomi.xmsf/stock/split-XiaomiServiceFrameworkCN-master/sources/pa/i.java
 * Stock class name is obfuscated as pa.i; this file keeps the deobfuscated com.xiaomi.slim.SlimConnection API.
 */
class SlimConnection(
    pushAction: IPushServiceAction,
    context: android.content.Context,
    connectionConfiguration: ConnectionConfiguration,
) : SocketConnection(pushAction, context, connectionConfiguration) {
    private var mReader: BlobReader? = null
    private var mReaderThread: Thread? = null
    private var mWriter: BlobWriter? = null
    private var mDerivedKey: ByteArray? = null
    @Volatile
    private var isShuttingDown = false

    private fun getPing(isServerPing: Boolean): Blob {
        val ping = Ping()
        if (isServerPing) {
            ping.packetID = "1"
        }
        val statsBytes = StatsHelper.retriveStatsAsByte()
        if (statsBytes != null) {
            val xMMsgPing = ChannelMessage.XMMsgPing().apply {
                setStats(com.google.protobuf.micro.ByteStringMicro.copyFrom(statsBytes))
            }
            @Suppress("DEPRECATION")
            ping.setPayload(xMMsgPing.toByteArray(), null)
        }
        return ping
    }

    @Throws(XMPPException::class)
    private fun initReaderAndWriter() {
        try {
            val sock = socket ?: throw XMPPException("Socket is null")
            mReader = BlobReader(sock.getInputStream(), this)
            mWriter = BlobWriter(sock.getOutputStream(), this)
            val threadName = "Blob Reader (${connectionCounterValue})"
            val thread = Thread(object : Runnable {
                override fun run() {
                    try {
                        mReader?.start()
                    } catch (e: Exception) {
                        notifyConnectionError(9, e)
                    }
                }
            }, threadName)
            mReaderThread = thread
            thread.start()
        } catch (e: Exception) {
            throw XMPPException("Error to init reader and writer", e)
        }
    }

    override fun batchSend(blobArray: Array<Blob>) {
        for (blob in blobArray) {
            send(blob)
        }
    }

    override fun batchSendPacket(packetArr: Array<Packet>) {
        for (packet in packetArr) {
            sendPacket(packet)
        }
    }

    @Throws(XMPPException::class)
    override fun bind(clientLoginInfo: PushClientsManager.ClientLoginInfo) {
        synchronized(this) {
            if (challenge.isNullOrEmpty()) {
                MyLog.w("[Slim] delay bind chid=${clientLoginInfo.chid} as challenge is missing")
                return
            }
            Binder.bind(clientLoginInfo, challenge, this)
        }
    }

    override val key: ByteArray?
        get() {
            synchronized(this) {
                if (mDerivedKey == null && !challenge.isNullOrEmpty()) {
                    val deviceUuid = ServiceConfig.getDeviceUUID()
                    if (!deviceUuid.isNullOrEmpty()) {
                        val challengeTail = challenge!!.substring(challenge!!.length / 2)
                        val deviceUuidTail = deviceUuid.substring(deviceUuid.length / 2)
                        mDerivedKey = RC4Cryption.encrypt(
                            challenge!!.toByteArray(),
                            (challengeTail + deviceUuidTail).toByteArray(),
                        )
                        if (mDerivedKey != null) {
                            mPushAction.runtimeObserver.onChannelEvent(null, "slim_key_derived", "SlimConnection.getKey")
                        }
                    }
                }
                return mDerivedKey
            }
        }

    @Throws(XMPPException::class, IOException::class)
    override fun initConnection() {
        synchronized(this) {
            initReaderAndWriter()
            mWriter?.openStream()
        }
    }

    override fun isBinaryConnection(): Boolean = true

    internal fun notifyDataArrived(blob: Blob?) {
        if (blob == null) return
        if (blob.hasErr()) {
            MyLog.w("[Slim] RCV blob chid=${blob.channelId}; id=${blob.packetID}; errCode=${blob.errCode}; err=${blob.errStr}")
        }
        val inboundPlan = mPushAction.runtimeObserver.resolveSlimInboundPlan(blob.channelId, blob.cmd)
        if (inboundPlan.action == PushSlimInboundAction.PingReceived) {
            MyLog.w("[Slim] RCV ping id=${blob.packetID}")
        }
        
        if (inboundPlan.action == PushSlimInboundAction.ChallengeReceived) {
            onChallengeReceived(extractChallenge(blob), "SlimConnection.notifyDataArrived")
        }
        
        inboundPlan.eventAction?.let { eventAction ->
            mPushAction.runtimeObserver.onChannelEvent(null, eventAction, "SlimConnection.notifyDataArrived")
        }
        inboundPlan.connectionState?.let { connectionState ->
            mPushAction.runtimeObserver.onConnectionStateChanged(
                connectionState.name,
                "SlimConnection.notifyDataArrived",
                host,
                inboundPlan.connectionReason.orEmpty()
            )
        }
        inboundPlan.disconnectReasonCode?.let { disconnectReasonCode ->
            notifyConnectionError(disconnectReasonCode, null)
        }
    }

    override fun notifyConnectionError(reason: Int, exc: Exception?) {
        if (isShuttingDown) return
        mPushAction.executeJob(object : com.xiaomi.push.service.XMPushServiceJob(2) {
            override fun getDesc(): String = "shutdown the connection. $reason, $exc"
            override fun process() {
                mPushAction.disconnect(reason, exc)
            }
        })
    }

    internal fun notifyDataArrived(packet: Packet?) {
        if (packet == null) return
        super.notifyDataArrived(packet)
    }

    private fun extractChallenge(blob: Blob): String? {
        if (blob.cmd == Blob.CMD_CONN) {
            return runCatching {
                ChannelMessage.XMMsgConnResp.parseFrom(blob.payload).challenge
            }.getOrNull()
        }
        return runCatching { String(blob.payload) }.getOrNull()
    }

    internal fun onChallengeReceived(receivedChallenge: String?, source: String) {
        val shouldRebind = synchronized(this) {
            if (receivedChallenge.isNullOrEmpty()) {
                MyLog.w("[Slim] RCV challenge missing in CONN response from $source")
                return
            }
            if (challenge == receivedChallenge) {
                return
            }
            challenge = receivedChallenge
            mDerivedKey = null
            true
        }
        MyLog.w("[Slim] RCV challenge=$receivedChallenge")
        if (shouldRebind) {
            rebindClientsAfterChallenge()
        }
    }

    private fun rebindClientsAfterChallenge() {
        mPushAction.executeJob(object : com.xiaomi.push.service.XMPushServiceJob(0) {
            override fun getDesc(): String = "re-bind after challenge"
            override fun process() {
                PushClientsManager.getInstance().getAllClients().forEach { client ->
                    if (client.status == PushClientsManager.ClientStatus.unbind || client.status == PushClientsManager.ClientStatus.binding) {
                        MyLog.w("[Slim] auto bind chid=${client.chid} after challenge")
                        mPushAction.executeJob(com.xiaomi.push.service.BindJob(mPushAction, client))
                    }
                }
            }
        })
    }

    override fun send(blob: Blob) {
        if (isShuttingDown) {
            MyLog.w("[Slim] skip sending blob as connection is shutting down")
            return
        }
        val writer = mWriter
        if (writer == null) {
            notifyConnectionError(10, IOException("the writer is null."))
            return
        }
        try {
            val bytesWritten = writer.write(blob)
            setReadAlive() // Corrected: should be write alive but matching current Connection.kt simplicity
            val packageName = blob.packageName
            if (!TextUtils.isEmpty(packageName)) {
                try {
                    TrafficUtils.distributionTraffic(
                        mContext,
                        packageName!!,
                        bytesWritten.toLong(),
                        false,
                        true,
                        System.currentTimeMillis(),
                    )
                } catch (th: Throwable) {
                    MyLog.e(th)
                }
            }
            notifyPacketSent(blob)
        } catch (e: Exception) {
            MyLog.w("[Slim] send blob failed: $e")
            notifyConnectionError(10, e)
        }
    }

    override fun sendPacket(packet: Packet) {
        @Suppress("DEPRECATION")
        send(Blob.from(packet, null))
    }

    @Throws(XMPPException::class)
    override fun sendPing(isServerPing: Boolean) {
        val ping = getPing(isServerPing)
        val pingPlan = mPushAction.runtimeObserver.resolveSlimSendPingPlan()
        MyLog.w("[Slim] SND ping id=${ping.packetID}")
        mPushAction.runtimeObserver.onChannelEvent(null, pingPlan.eventAction, "SlimConnection.sendPing")
        send(ping)
    }

    override fun shutdown(
        reason: Int,
        error: Exception?,
    ) {
        synchronized(this) {
            if (isShuttingDown) return
            isShuttingDown = true
            mReader?.let {
                it.shutdown()
                mReader = null
            }
            mWriter?.let {
                try {
                    it.shutdown()
                } catch (e: Exception) {
                    MyLog.e(e)
                }
                mWriter = null
                mDerivedKey = null
                super.shutdown(reason, error)
            } ?: run {
                mDerivedKey = null
                super.shutdown(reason, error)
            }
        }
    }

    @Throws(XMPPException::class)
    override fun unbind(
        chid: String,
        userId: String,
    ) {
        synchronized(this) {
            Binder.unbind(chid, userId, this)
        }
    }
}
